[Concurrency] Don't pass negative times to the Dispatch code.

al45tair opened the pull request:
Dispatch uses unsigned times, and cannot cope with times before its clock started.  As such, passing negative times from Swift through to the C++ code results in large unsigned values, which then causes us to wait forever.  This is undesirable.

rdar://148899609

al45tair committed:
[Concurrency] Don't pass negative times to the Dispatch code.

Dispatch uses unsigned times, and cannot cope with times before its
clock started.  As such, passing negative times from Swift through to
the C++ code results in large unsigned values, which then causes us to
wait forever.  This is undesirable.

rdar://148899609

al45tair commented:
I'm going to add a test for this as well.

weissi commented on stdlib/public/Concurrency/DispatchExecutor.swift:
```diff
@@ -161,20 +161,29 @@ enum DispatchClockID: CInt {
 
+  func clamp(_ components: (seconds: Int64, attoseconds: Int64))
+    -> (seconds: Int64, attoseconds: Int64) {
+    if components.seconds < 0
```
What happens if `components.seconds == 1` but `components.attoseconds` is so negative that the compound is still less than 0?

`(seconds: 1, attoseconds: -2_000_000_000_000_000_000)` or so?

weissi replied:
> (seconds: 1, attoseconds: -2_000_000_000_000_000_000)

or ofc `(seconds: 1, attoseconds: Int64.min)` which I believe is -9.something seconds

weissi replied:
```
@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
extension Duration {

    /// The composite components of the `Duration`.
    ///
    /// This is intended for facilitating conversions to existing time types. The
    /// attoseconds value will not exceed 1e18 or be lower than -1e18.
    @available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
    public var components: (seconds: Int64, attoseconds: Int64) { get }
}
```

A user can just set those fields independently, right?

al45tair replied:
No. If you try to use the `init()` that takes those components, you'll find it adds and multiplies in order to get a count of just attoseconds. And `components` is `get`-only.

weissi replied:
Ah I see, thank you!

al45tair committed:
[Concurrency][Tests] Add a test of waiting until a time before the epoch.

Add a test to check that attempting to `Task.sleep(until:)` passing a
time before the start of the Dispatch clocks works and doesn't wait
forever.

rdar://148899609

al45tair commented:
@swift-ci Please smoke test
